src/mame/includes/tsconf.h

```diff
@@ -0,0 +1,212 @@
+#include "spec128.h"
+#include "machine/tsconfdma.h"
+#include "machine/nvram.h"
+#include "machine/pckeybrd.h"
+#include "machine/8042kbdc.h"
+#include "tilemap.h"
+
+#define LOG_ERROR (1U << 0)
+#define LOG_WARN (1U << 1)
+#define LOG_DEBUG (1U << 2)
+
+#define VERBOSE (LOG_WARN)
+#define LOG_OUTPUT_FUNC osd_printf_info
+
+#include "logmacro.h"
+#define LOGERROR(...) LOGMASKED(LOG_DEBUG | LOG_WARN | LOG_ERROR, __VA_ARGS__)
+#define LOGWARN(...) LOGMASKED(LOG_DEBUG | LOG_WARN, __VA_ARGS__)
+#define LOGDEBUG(...) LOGMASKED(LOG_DEBUG, __VA_ARGS__)
+
+#define PAGE(_r) ((_r) << 14)
+#define RAM_PAGE_OFFST(_page, _offset) (PAGE(_page) + _offset)
+
+#define REG(_p) m_regs[_p]
+#define REGNUM(_r) (&(_r) - &(V_CONFIG))
+
+#define V_CONFIG REG(0x00)
+#define V_PAGE REG(0x01)
+
+#define G_X_OFFS_L REG(0x02)
+#define G_X_OFFS_H REG(0x03)
+#define G_Y_OFFS_L REG(0x04)
+#define G_Y_OFFS_H REG(0x05)
+
+#define TS_CONFIG REG(0x06)
+#define PAL_SEL REG(0x07)
+#define BORDER REG(0x0f)
+
+#define PAGE0 REG(0x10)
+#define PAGE1 REG(0x11)
+#define PAGE2 REG(0x12)
+#define PAGE3 REG(0x13)
+#define FMAPS REG(0x15)
+
+#define T_MAP_PAGE REG(0x16)
+#define T0_G_PAGE REG(0x17)
+#define T1_G_PAGE REG(0x18)
+#define SG_PAGE REG(0x19)
+
+#define DMAS_ADDRESS_L REG(0x1a)
+#define DMAS_ADDRESS_H REG(0x1b)
+#define DMAS_ADDRESS_X REG(0x1c)
+#define DMAD_ADDRESS_L REG(0x1d)
+#define DMAD_ADDRESS_H REG(0x1e)
+#define DMAD_ADDRESS_X REG(0x1f)
+#define DMA_WAIT_PORT_DEV REG(0x25)
+#define DMA_LEN REG(0x26)
+#define DMA_CTRL REG(0x27)
+#define DMA_NUM_L REG(0x28)
+#define DMA_NUM_H REG(0x2c)
+
+#define SYS_CONFIG REG(0x20)
+#define MEM_CONFIG REG(0x21)
+#define HS_INT REG(0x22)
+#define VS_INT_L REG(0x23)
+#define VS_INT_H REG(0x24)
+
+#define INT_MASK REG(0x2a)
+
+#define T0_X_OFFSER_L REG(0x40)
+#define T0_X_OFFSER_H REG(0x41)
+#define T0_Y_OFFSER_L REG(0x42)
+#define T0_Y_OFFSER_H REG(0x43)
+#define T1_X_OFFSER_L REG(0x44)
+#define T1_X_OFFSER_H REG(0x45)
+#define T1_Y_OFFSER_L REG(0x46)
+#define T1_Y_OFFSER_H REG(0x47)
+
+static constexpr rgb_t rgb15(u16 rgb)
+{
+	u8 r = (BIT(rgb, 10, 5) << 3) | 0x07;
+	u8 g = (BIT(rgb, 5, 5) << 3) | 0x07;
+	u8 b = (BIT(rgb, 0, 5) << 3) | 0x07;
+	return rgb_t(r, g, b);
+}
+
+class tsconf_state : public spectrum_128_state
+{
+public:
+	tsconf_state(const machine_config &mconfig, device_type type, const char *tag)
+		: spectrum_128_state(mconfig, type, tag),
+		  m_bank1(*this, "bank1"), m_bank2(*this, "bank2"), m_bank3(*this, "bank3"), m_bank4(*this, "bank4"),
+		  m_keyboard(*this, "pc_keyboard"),
+		  m_beta(*this, BETA_DISK_TAG),
+		  m_sdcard(*this, "sdcard"),
+		  m_dma(*this, "dma"), m_cmos(*this, "cmos"), m_glukrs(*this, "glukrs"),
+		  m_gfxdecode(*this, "gfxdecode"),
+		  m_cram(*this, "cram")
+	{
+	}
+
+	void tsconf(machine_config &config);
+
+protected:
+	virtual void video_start() override;
+	virtual void machine_start() override;
+	virtual void machine_reset() override;
+
+private:
+	DECLARE_VIDEO_START(tsconf);
+
+	enum gluk_ext
```

Comment:
If you want to be able to register instances of this for save states, you need to declare an explicit storage class.